plugin/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala

```diff
@@ -146,78 +163,89 @@ object ScalafmtPlugin extends AutoPlugin {
   }
 
   private def checkSources(
-      cacheDirectory: File,
+      cacheStoreFactory: CacheStoreFactory,
       sources: Seq[File],
       config: Path,
       log: Logger,
       writer: PrintWriter
-  ): Boolean = {
-    cached(cacheDirectory, FilesInfo.lastModified, config) { cacheMisses =>
-      val changed = cacheMisses.filter(_.exists)
-      if (changed.size > 0) {
-        log.info(s"Checking ${changed.size} Scala sources...")
-        checkSources(changed.toSeq, config, log, writer)
-      } else {
-        true
-      }
-    }(sources.toSet).getOrElse(true)
+  ): ScalafmtAnalysis = {
+    trackSourcesAndConfig(cacheStoreFactory, sources, config) {
+      (outDiff, configChanged, prev) =>
+        log.debug(outDiff.toString)
+        val updatedOrAdded = outDiff.modified & outDiff.checked
+        val filesToCheck =
+          if (configChanged) sources
+          else updatedOrAdded.toList
+        val failed = prev.failedScalafmtCheck filter { _.exists }
+        val files: List[File] = (filesToCheck.toList ++ failed).distinct
```

Comment:
Extending incrementality to rejected/failed sources (and not just formatted ones) is probably beyond the scope of https://github.com/scalameta/sbt-scalafmt/issues/42, but I guess we don't really need to recheck the files that failed the previous run - running a fresh check against `filesToCheck` only and concatenating the result with the previous analysis (intersected with `outDiff.unmodified`) should do the trick, no? This would require to push up the `log.warn(s"${file.toString} isn't formatted properly!")` though if we want to keep logging the list of files that need to be formatted, no matter the status of the cache.